tests/numerics/nonlinear_solver_selector_02.cc

```diff
@@ -0,0 +1,408 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2007 - 2023 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+
+// Tests the class NonlinearSolverSelector using an example based on
+// the step-77 tutorial. The output will vary depending on what
+// packages are configured with deal.II.
+
+#include <deal.II/base/function.h>
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/base/timer.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_q.h>
+#include <deal.II/fe/fe_values.h>
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/grid_refinement.h>
+#include <deal.II/grid/tria.h>
+
+#include <deal.II/lac/affine_constraints.h>
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/sparse_direct.h>
+#include <deal.II/lac/sparse_matrix.h>
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/error_estimator.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/nonlinear.h>
+#include <deal.II/numerics/solution_transfer.h>
+#include <deal.II/numerics/vector_tools.h>
+#ifdef DEAL_II_WITH_SUNDIALS
+#  include <deal.II/sundials/kinsol.h>
+#endif
+#ifdef DEAL_II_WITH_TRILINOS
+#  include <deal.II/trilinos/nox.h>
+#endif
+
+//#include "nonlinear_solver_selector_01.cc"
+
+#include <fstream>
+#include <iostream>
+
+#include "../tests.h"
+
+
+namespace nonlinear_solver_selector_test
+{
+  using namespace dealii;
+
+  using NLSolve = NonlinearSolverSelector<Vector<double>>;
+
+#ifndef SOLVER
+#  define SOLVER NLSolve::AdditionalData::nox
+#endif
```

Comment:
This test duplicates ~400 lines of code. Why not follow the suggestion of @tjhei to only `#define` this macro and then `#include` the other `.cc` file? That seemed like a reasonable solution with a minimal amount of code duplication.